### PR TITLE
Cargo mutants version update

### PR DESCRIPTION
**Pull request by benalleng**
There are still 2 remaining mutants after all of the exclusions I still have todo.
```
Found 444 mutants to test
ok       Unmutated baseline in 32.1s build + 11.9s test
 INFO Auto-set test timeout to 1m
MISSED   payjoin/src/core/receive/v1/mod.rs:519:33: replace > with >= in WantsInputs::avoid_uih in 3.4s build + 28.9s test
444 mutants tested in 13m 52s: 1 missed, 280 caught, 163 unviable
```

The exclusions are catogerized as follows.

- expiry time checks I found not worth creating a test for.
- mutations that allow code to simply skip or follow a code path that does not actually cause or prevent an error
- The non deterministic mutation

### CODE REVIEW

File: payjoin/src/core/receive/v1/mod.rs [hunk 1]
```diff
@@ -933,6 +933,16 @@ pub(crate) mod test {
...
             _ => panic!("Broadcast suitability check should fail due to being below the min fee rate or unexpected error type"),
         };
+
+        let min_fee_rate = proposal.psbt_fee_rate().expect("Feerate calculation should not fail");
+        match proposal.clone().check_broadcast_suitability(Some(min_fee_rate), |_| Ok(true)) {
+            Ok(_) => {
+                assert_eq!(proposal.clone().psbt_fee_rate().unwrap(), min_fee_rate)
+            }
+            Err(_) => {
+                panic!("Broadcast suitability check should fail due to being below the min fee rate or unexpected error type")
```
Comment (spacebear21) [inline payjoin/src/core/receive/v1/mod.rs]:
```suggestion
                panic!("Broadcast suitability check with appropriate min_fee_rate should succeed")
```

Comment (spacebear21) [reply payjoin/src/core/receive/v1/mod.rs]: Actually, I'd probably re-write the whole match statement:

```rust
let min_fee_rate = proposal.psbt_fee_rate().expect("Feerate calculation should not fail");
let _ = proposal.clone().check_broadcast_suitability(Some(min_fee_rate), |_| Ok(true)).expect("Broadcast suitability check with appropriate min_fee_rate should succeed");
assert_eq!(proposal.clone().psbt_fee_rate().unwrap(), min_fee_rate);
```

Re-writing it like that, it's unclear what exactly this assert_eq is testing for? It seems tautological that min_fee_rate == `proposal.psbt_fee_rate().unwrap()`

File: payjoin/src/core/receive/v1/mod.rs [hunk 2]
```diff
@@ -933,6 +933,16 @@ pub(crate) mod test {
             },
             _ => panic!("Broadcast suitability check should fail due to being below the min fee rate or unexpected error type"),
```
Comment (spacebear21) [inline payjoin/src/core/receive/v1/mod.rs]: Probably this match could also be re-written as an `expect_err()` instead of match/panic.

File: payjoin/src/core/receive/v1/mod.rs [hunk 3]
```diff
@@ -1140,6 +1150,17 @@ pub(crate) mod test {
...
             .script_pubkey;
 
+        let output_value =
+            wants_outputs.original_psbt.unsigned_tx.output[wants_outputs.change_vout].value;
+        let outputs = vec![TxOut { value: output_value, script_pubkey: script_pubkey.clone() }];
+        let increased_amount =
+            wants_outputs.clone().replace_receiver_outputs(outputs, script_pubkey.as_script());
+        assert!(
+            increased_amount.is_ok(),
+            "Not changing the receiver output amount should always be allowed"
```
Comment (spacebear21) [inline payjoin/src/core/receive/v1/mod.rs]:
```suggestion
            "Increasing the receiver output amount is always allowed"
```

File: payjoin/src/core/send/mod.rs [hunk 4]
```diff
@@ -543,6 +543,15 @@ mod test {
             fee_contribution.err(),
             Some(InternalBuildSenderError::FeeOutputValueLowerThanFeeContribution)
         );
+        let fee_contribution = determine_fee_contribution(
```
Comment (spacebear21) [inline payjoin/src/core/send/mod.rs:776]: Some explanatory comments on these new edge cases wouldn't hurt. I'm guessing this is testing the maximum allowed fee contribution based on the input value?

File: payjoin/src/core/send/v1.rs [hunk 5]
```diff
@@ -480,10 +480,12 @@ mod test {
     #[test]
     fn process_response_invalid_utf8() {
         // In UTF-8, 0xF0 represents the start of a 4-byte sequence, so 0xF0 by itself is invalid
```
Comment (spacebear21) [inline payjoin/src/core/send/v1.rs]: This comment looks obsolete now.

File: .cargo/mutants.toml [hunk 6]
```diff
@@ -9,7 +10,24 @@ exclude_re = [
...
+	"replace < with <= in SenderBuilder<'a>::build_recommended", # clamping the fee contribution when the fee equals to the recommended fee does not do anything
+
+	# Async SystemTime comparison
+	"replace > with >= in Sender<WithReplyKey>::extract_v2", # checking if the system time is equal to the expiry is difficult to reasonably test
+	"replace < with <= in Receiver<Initialized>::apply_unchecked_from_payload", # checking if the system time is equal to the expiry is difficult to reasonably test 
+	"replace > with >= in Receiver<Initialized>::extract_req", # checking if the system time is equal to the expiry is difficult to reasonably test
+	"replace > with >= in extract_err_req", # checking if the system time is equal to the expiry is difficult to reasonably test
+        "replace > with >= in Receiver<Initialized>::create_poll_request", # checking if the system time is equal to the expiry is difficult to reasonably test
+	"replace > with >= in Sender<WithReplyKey>::create_v2_post_request", # checking if the system time is equal to the expiry is difficult to reasonably test
```
Comment (spacebear21) [inline .cargo/mutants.toml]: nit: 

```suggestion
	# Async SystemTime comparison
	# checking if the system time is equal to the expiry is difficult to reasonably test
	"replace > with >= in Sender<WithReplyKey>::extract_v2",
	"replace < with <= in Receiver<Initialized>::apply_unchecked_from_payload",
	"replace > with >= in Receiver<Initialized>::extract_req",
	"replace > with >= in extract_err_req",
        "replace > with >= in Receiver<Initialized>::create_poll_request",
	"replace > with >= in Sender<WithReplyKey>::create_v2_post_request",
```

File: .cargo/mutants.toml [hunk 7]
```diff
@@ -9,7 +10,24 @@ exclude_re = [
...
+	# Timeout loops
         "interleave_shuffle", # Replacing index += 1 with index *= 1 in a loop causes a timeout due to an infinite loop
+
+	# Trivial mutations
+	"replace > with >= in ProvisionalProposal::apply_fee", # allowing 2 code blocks in this function to run when the additional fee = 0 does nothing
+	"replace < with <= in PsbtContext::check_outputs", # allowing the subtraction of 2 equal values always resulting in a contrib_fee of 0
+	"replace > with >= in PsbtContext::check_fees", # checking if the feerate is below the minimum when the minimum is allowed to be zero does nothing
+	"replace < with <= in SenderBuilder<'a>::build_recommended", # clamping the fee contribution when the fee equals to the recommended fee does not do anything
+
+	# Async SystemTime comparison
```
Comment (spacebear21) [inline .cargo/mutants.toml:27]: We may want to figure out a way to mock SystemTime in tests. I've used [freezegun](https://pypi.org/project/freezegun/) in Python in the past, I wonder if a similar tool exists the Rust ecosystem. We may even be able to mock it with a Trait, per Claude:
```rust
use std::time::SystemTime;

trait TimeProvider {
    fn now(&self) -> SystemTime;
}

struct RealTimeProvider;

impl TimeProvider for RealTimeProvider {
    fn now(&self) -> SystemTime {
        SystemTime::now()
    }
}

struct MockTimeProvider {
    time: SystemTime,
}

impl TimeProvider for MockTimeProvider {
    fn now(&self) -> SystemTime {
        self.time
    }
}

// Your code uses the trait
fn process_with_timestamp<T: TimeProvider>(time_provider: &T) -> SystemTime {
    time_provider.now()
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::time::{Duration, UNIX_EPOCH};

    #[test]
    fn test_with_mock_time() {
        let mock_time = UNIX_EPOCH + Duration::from_secs(1000);
        let provider = MockTimeProvider { time: mock_time };
        
        let result = process_with_timestamp(&provider);
        assert_eq!(result, mock_time);
    }
}
```
This seems reasonable to leave as a follow-up, maybe to be addressed alongside #893.

**Review by spacebear21 — CHANGES_REQUESTED** [review]
I slightly dislike the mutants exclusions by name because they seem vulnerable to regressions (e.g. if the skipped functions get renamed as #884 is likely to do for some of these) or false positives (e.g. what was once a trivial mutation becomes serious after several refactors). Trivial mutations seem like a code smell that the underlying code could be re-written to eliminate ambiguity. I don't have specific suggestions regarding those excluded mutants, but I want to ensure that we're very deliberate about what we choose to exclude.

File: .cargo/mutants.toml (hunk 7 above)
Comment (benalleng) [reply .cargo/mutants.toml:27]: Ahh interesting, I got hung up on trying to fit the tokio::test pause() method in but it just wasn't working.

**Comment by benalleng** [comment]
Thanks! as for those trivial mutants a good can of axe body spray by just making the code blocks follow the logic path in the inverted operand cases. 

With the named exclusions I think we might want to consider using diff mutations https://github.com/sourcefrog/cargo-mutants/blob/1481918f377a2e0a86a66185cb536dd292157438/.github/workflows/tests.yml#L234-L277 perhaps we're not quite ready for it in our codebase but it would be nice to keep the mutant coverage more maintainable

File: .cargo/mutants.toml (hunk 6 above)
Comment (benalleng) [reply .cargo/mutants.toml]: Tried to to a little further organization

**Review by spacebear21 — APPROVED** [review]
utACK 7a21ab73e9870d6935d430a3bf4a3c7652bec33c

It would be great to address the exclusions in later follow-ups and remove them from the exclusion list, but these seem reasonable enough for now, and merging this PR will reduce the noise from automated mutants issues.
